crates/gpui/src/platform/windows/platform.rs

```diff
@@ -250,36 +225,38 @@ impl Platform for WindowsPlatform {
 
     fn run(&self, on_finish_launching: Box<dyn 'static + FnOnce()>) {
         on_finish_launching();
-        'a: loop {
-            let mut msg = MSG::default();
-            // will be 0 if woken up by self.inner.event or 1 if the compositor clock ticked
-            // SEE: https://learn.microsoft.com/en-us/windows/win32/directcomp/compositor-clock/compositor-clock
-            let wait_result =
-                unsafe { DCompositionWaitForCompositorClock(Some(&[self.inner.event]), INFINITE) };
-
-            // compositor clock ticked so we should draw a frame
-            if wait_result == 1 {
-                unsafe {
-                    invalidate_thread_windows(
-                        GetCurrentThreadId(),
-                        &self.inner.window_handle_values.borrow(),
-                    )
-                };
-
-                while unsafe { PeekMessageW(&mut msg, HWND::default(), 0, 0, PM_REMOVE) }.as_bool()
-                {
-                    if msg.message == WM_QUIT {
-                        break 'a;
-                    }
+        let tick_event = unsafe { CreateEventW(None, false, false, None).unwrap() };
+        let dispatch_event = self.inner.event;
+        generate_compositor_tick(tick_event);
 
-                    if !self.run_immediate_msg_handlers(&msg) {
-                        unsafe { TranslateMessage(&msg) };
-                        unsafe { DispatchMessageW(&msg) };
+        'a: loop {
+            let event = unsafe {
+                MsgWaitForMultipleObjects(
+                    Some(&[tick_event, dispatch_event]),
+                    false,
+                    INFINITE,
+                    QS_ALLINPUT,
+                )
```

Comment:
This is not necessary as far as I know. `DCompositionWaitForCompositorClock` can be woken up by events as well. A good document for it is available here. https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src/directcomp/compositor-clock/compositor-clock.md

Reply:
In view of the fallback in Windows 10 to use a different timer, I think it would be more flexible for changes to use a `tick_event` with an independent timer thread.

Reply:
maybe, but using `DCompositionWaitForCompositorClock` inside a thread pool only to trigger a `MsgWaitForMultipleObjects` seems odd to me. We're waiting on 2 mechanisms which are using 4 different handlers (2 from us, 2 from the system). Seems excessive.

Reply:
It seems that `DCompositionWaitForCompositorClock` is only awakened by user-defined events, in this case, `dispatch_event`, so, for example, the user's keyboard input is actually blocked. I guess I haven't misunderstood this. In my modification, the user's keyboard input will be processed and rendered immediately. Additionally, Microsoft advises that "applications are expected to use this API to control rendering threads, separate from general-purpose UI threads," so I assume that using the default thread pool provided by Windows should consume minimal resources.

Reply:
Yes, it's woken up by user events and the clock. Window events don't need to be processed until the compositor tick is ticked with how GPUI is designed since it won't matter unless a render occurs and that render is supposed to happen at the displays refresh rate.

My question is does your change to the loop make a difference in some regard? If it does then I'd love to hear it. Did it not work? I don't think we should change the way the loop works unless there is a clear benefit to GPUI.

As it stands now it's creating an extra worker thread and 2 synchronization points (which whatever primitive windows uses under the hood.) Before it was just 1 synchronization point.

The only argument I see with this change right now is the flexibility kazatsuyu mentioned for adding a timer for win10, but I want to stress again - is this making a meaningful difference? If it does then great!

Reply:
> Yes, it's woken up by user events and the clock. Window events don't need to be processed until the compositor tick is ticked with how GPUI is designed since it won't matter unless a render occurs and that render is supposed to happen at the displays refresh rate.

Makes sense.